kolibri/core/assets/src/components.js

```diff
@@ -149,6 +73,132 @@ var CrudCollection = Mn.CollectionView.extend({
+    // In other words, if the view instance is given a model attribute when instantiated, then the argument
+    // to this function will be a copy of that model's .attributes hash.
+    template: function() {
+        var html =
+            '{{#unless valid}}<span>Invalid Input!</span><br />{{/unless}}' +
+            '{{#each create}}' +
+                '<input data-attr="{{ name }}" type="text" placeholder="{{ name }}" value="{{ default }}" /><br />' +
+            '{{/each}}' +
+            '<button class="create flat-button">Create</button>';
+        return Handlebars.compile(html)({create: this.create, valid: this.valid});
+    },
+
+    triggers: {
+        // Re-triggers the specified DOM event (from its template) as a Backbone.Event on this view.
+        'click .create': 'create'
+    },
+
+    initialize: function(options) {
+        // Some options when passed to the constructor are automatically bound to the view instance.
+        // This is the case for the "model" attribute, so passing a model to a view constructor means that the
+        // instance has it available automatically as "this.model".
+        // Custom options (like "create" below) must be bound to the view instance manually, or otherwise handled
+        // by the initialize function.
+        this.create = _.map(options.create, function(attr) {
+            return {
+                name: attr,
+                default: ''
+            };
+        });
+        this.validators = options.validators;
+        this.valid = true;
+        // Since "template" is a callback, bindAll ensures "this" refers the view instance.
+        _.bindAll(this, 'template');
+    },
+
+    // For event "xyz", you can specify a handler without having to set up a listener using the camel-cased
+    // "onXyz" convention.
+    // E.g. "create" event is handled automatically by the "onCreate" function if it exists.
+    onCreate: function() {
+        if ( this.validate() ) {
+            _.forEach(this.create, _.bind(function (obj) {
+                var attr = obj.name;
+                this.model.set(attr, this.getAttrVal(attr));
+            }, this));
+            this.collection.add(this.model);
+            this.trigger('closeModal');
+        } else {
+            // Shows an "Invalid!" message.
+            this.rerender = true;
+            this.render();
+        }
+    },
+
+    onBeforeRender: function() {
+        // Before re-rendering, make sure that values user has already entered aren't trashed
+        if (this.rerender) {
+            var self = this;
+            this.create = _.map(this.create, function(old) {
+                return {
+                    name: old.name,
+                    default: self.getAttrVal(old.name)
+                };
+            });
+        }
+    },
+
+    getAttrVal: function(attr) {
+        var el = this.$el.find('input[data-attr=' + attr + ']');
+        return $(el).val();
+    },
```

Comment:
That's true - many view components need to have local state when that state is irrelevant to the application's global state. Validation state is a common example - I did the same thing by embedding validation state in `tag.validation` in the [riot.js add-user form](https://github.com/indirectlylit/kolibri/blob/riot-toy-app/kolibri/core/assets/src/riot-tags/usermgmt-addform.tag.html).

Here's an [interesting explanation of best practices](https://facebook.github.io/react/docs/interactivity-and-dynamic-uis.html#what-components-should-have-state) regarding React's `props`, which (if I understand correctly) are like options passed in from a parent or global app state, and `state`, which is a view's own internal UI state.